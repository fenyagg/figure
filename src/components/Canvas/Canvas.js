import { observer } from 'mobx-react-lite';
import React, {useContext} from 'react';
import './Canvas.scss';
import { StoreContext } from '../../stores';

const Canvas = () => {
  const context = useContext(StoreContext);
  return (
    <div className='canvas'>

    </div>
  );
};

export default observer(Canvas);
