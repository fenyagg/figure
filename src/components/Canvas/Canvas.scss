.canvas {
	background-color: #fff;
	border: 1px solid;
	position: relative;

	&__figure-container {
		position: absolute;

        &._active {
          outline: 1px solid red;
          z-index: 1;
          cursor: grab;

          &._dragging {
            cursor: grabbing;
		  	opacity: .5;
          }
        }
	}

	&__figure {
		position: relative;
		fill: #FF9800;
		stroke: black;
		stroke-linejoin: round;
		width: 100%;
		height: 100%;
	}
}
