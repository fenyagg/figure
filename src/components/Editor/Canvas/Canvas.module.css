.canvas {
	background-color: #fff;
	border: 1px solid;
	position: relative;
}

.cursorNW {
	cursor: nw-resize;
}

.cursorNE {
	cursor: nw-resize;
}
