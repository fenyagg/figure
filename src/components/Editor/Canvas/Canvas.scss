.canvas {
	background-color: #fff;
	border: 1px solid;
	position: relative;
}

