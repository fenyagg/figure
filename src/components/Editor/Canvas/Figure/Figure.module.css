.figure {
  position: absolute;
}

.figureImg {
  position: relative;
  fill: #FF9800;
  stroke: black;
  stroke-linejoin: round;
  width: 100%;
  height: 100%;
}

.figureSelected .figureImg {
  stroke: #86190e;
}
