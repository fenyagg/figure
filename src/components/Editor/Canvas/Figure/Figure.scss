$dotOffset: -7px;

.figure {
  position: absolute;

  &._active {
    outline: 1px solid red;
    z-index: 1;
    cursor: grab;

    &._dragging {
      cursor: grabbing;
      opacity: .5;
    }
  }

  &__img {
    position: relative;
    fill: #FF9800;
    stroke: black;
    stroke-linejoin: round;
    width: 100%;
    height: 100%;
  }

  &__dot {
    position: absolute;
    width: 14px;
    height: 14px;
    background-color: red;
    z-index: 1;
    display: none;

    .figure._active & {
      display: block;
    }

    &._left-top {
      cursor: nw-resize;
      left: $dotOffset;
      top: $dotOffset;
    }
    &._left-bot {
      cursor: ne-resize;
      left: $dotOffset;
      bottom: $dotOffset;
    }
    &._right-top {
      cursor: ne-resize;
      right: $dotOffset;
      top: $dotOffset;
    }
    &._right-bot {
      cursor: nw-resize;
      right: $dotOffset;
      bottom: $dotOffset;
    }
  }
}