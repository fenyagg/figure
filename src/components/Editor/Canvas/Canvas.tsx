import { observer } from 'mobx-react-lite';
import React, { MouseEvent, useContext, useRef } from 'react';
import { StoreContext } from '../../../stores';
import './Canvas.scss';
import Figure from './Figure/Figure';

const Canvas = () => {
  const context = useContext(StoreContext);
  const canvasRef = useRef(null);
  const onMouseDown = (e: MouseEvent) => {
    if (e.currentTarget === canvasRef.current) {
      context.canvas.setActiveFigure(null);
    }
  };
  const onMouseLeave = (e: MouseEvent) => {
    if (e.currentTarget === canvasRef.current) {
      context.canvas.setIsDragging(false);
    }
  };

  return (
    <div
      className={'canvas'}
      style={{ width: context.canvas.width, height: context.canvas.height }}
      onMouseLeave={onMouseLeave}
      onMouseDown={onMouseDown}
      ref={canvasRef}
    >
      {context.canvas.figures.map(figure => (
        <Figure key={figure.id} figure={figure} />
      ))}
    </div>
  );
};

export default observer(Canvas);
