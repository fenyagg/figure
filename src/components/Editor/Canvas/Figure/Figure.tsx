import classNames from 'classnames';
import { observer } from 'mobx-react-lite';
import React, { MouseEvent, useContext, useMemo } from 'react';
import { StoreContext } from 'stores';
import { IFigure } from '../../../../stores/Canvas';
import { EResizeType } from '../../../../stores/canvas.types';
import { figuresList } from '../../../svg';
import './Figure.scss';

interface IProps {
  figure: IFigure;
}

const dotList = [
  {
    type: EResizeType.LEFT_TOP,
    className: '_left-top',
  },
  {
    type: EResizeType.LEFT_BOT,
    className: '_left-bot',
  },
  {
    type: EResizeType.RIGHT_TOP,
    className: '_right-top',
  },
  {
    type: EResizeType.RIGHT_BOT,
    className: '_right-bot',
  },
];

const Figure: React.FC<IProps> = ({ figure }) => {
  const context = useContext(StoreContext);

  const isActive = useMemo(() => {
    return figure.id === context.canvas.activeFigureId;
  }, [figure.id, context.canvas.activeFigureId]);

  const onFigureClick = () => {
    if (!isActive) {
      context.canvas.setActiveFigure(figure.id);
    }
  };
  const onFigureMouseDown = (e: MouseEvent) => {
    e.stopPropagation();
    const isLeftMouseButton = e.button === 0;
    if (isLeftMouseButton && isActive) {
      context.canvas.setIsDragging(true);
      context.canvas.setDragPosition(e.pageX, e.pageY);
    }
  };
  const onDotMouseDown = (e: MouseEvent, resizeType: EResizeType) => {
    e.stopPropagation();
    const isLeftMouseButton = e.button === 0;
    if (isLeftMouseButton && isActive) {
      context.canvas.setResizingType(resizeType);
      context.canvas.setDragPosition(e.pageX, e.pageY);
    }
  };

  const FigureImg = figuresList[figure.type];
  return (
    <div
      key={figure.id}
      style={{
        width: figure.width,
        height: figure.height,
        left: figure.positionLeft,
        top: figure.positionTop,
      }}
      className={classNames('figure', {
        _active: isActive,
        _dragging: isActive && context.canvas.isDragging,
      })}
      onClick={onFigureClick}
      onMouseDown={onFigureMouseDown}
    >
      {dotList.map(dot => (
        <div
          key={dot.type}
          onMouseDown={e => onDotMouseDown(e, dot.type)}
          className={`figure__dot ${dot.className}`}
        />
      ))}

      <FigureImg className="figure__img" />
    </div>
  );
};

export default observer(Figure);
