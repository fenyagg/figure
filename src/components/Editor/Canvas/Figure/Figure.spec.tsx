import React from 'react';
import { render } from 'enzyme';
import { IFigure } from 'stores/models/Canvas/Canvas';
import { EFigureType } from 'stores/models/Canvas/canvas.types';
import Figure from './Figure';

it('snap', () => {
  const figureItem: IFigure = {
    id: 'test',
    type: EFigureType.CIRCLE,
    height: 0,
    width: 0,
    left: 0,
    top: 0,
  };
  const component = render(<Figure figure={figureItem} />);
  expect(component).toMatchSnapshot();
});
