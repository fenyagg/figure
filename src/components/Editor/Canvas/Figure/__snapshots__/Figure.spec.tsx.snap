// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "figure",
      "style": "width:0;height:0;transform:translate3d(
          0px,
          0px,
          0
        )",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "figure__dot _left-top",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "figure__dot _left-bot",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "figure__dot _right-top",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "figure__dot _right-bot",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "figure__img",
                  "preserveAspectRatio": "none",
                  "viewBox": "-1 -1 32 32",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "cx": "15",
                      "cy": "15",
                      "rx": "15",
                      "ry": "15",
                    },
                    "children": Array [],
                    "name": "ellipse",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "cx": undefined,
                      "cy": undefined,
                      "rx": undefined,
                      "ry": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "cx": undefined,
                      "cy": undefined,
                      "rx": undefined,
                      "ry": undefined,
                    },
                  },
                ],
                "name": "svg",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "preserveAspectRatio": undefined,
                  "viewBox": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "preserveAspectRatio": undefined,
                  "viewBox": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "figure__dot _left-bot",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "figure__dot _right-top",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "figure__dot _right-bot",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "figure__img",
                "preserveAspectRatio": "none",
                "viewBox": "-1 -1 32 32",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "cx": "15",
                    "cy": "15",
                    "rx": "15",
                    "ry": "15",
                  },
                  "children": Array [],
                  "name": "ellipse",
                  "namespace": "http://www.w3.org/2000/svg",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "cx": undefined,
                    "cy": undefined,
                    "rx": undefined,
                    "ry": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "cx": undefined,
                    "cy": undefined,
                    "rx": undefined,
                    "ry": undefined,
                  },
                },
              ],
              "name": "svg",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "preserveAspectRatio": undefined,
                "viewBox": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "preserveAspectRatio": undefined,
                "viewBox": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "figure__dot _left-top",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "figure__dot _right-top",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "figure__dot _right-bot",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "figure__img",
              "preserveAspectRatio": "none",
              "viewBox": "-1 -1 32 32",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "cx": "15",
                  "cy": "15",
                  "rx": "15",
                  "ry": "15",
                },
                "children": Array [],
                "name": "ellipse",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "cx": undefined,
                  "cy": undefined,
                  "rx": undefined,
                  "ry": undefined,
                },
                "x-attribsPrefix": Object {
                  "cx": undefined,
                  "cy": undefined,
                  "rx": undefined,
                  "ry": undefined,
                },
              },
            ],
            "name": "svg",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "preserveAspectRatio": undefined,
              "viewBox": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "preserveAspectRatio": undefined,
              "viewBox": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "figure__dot _left-bot",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "figure__dot _left-top",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "figure__dot _right-bot",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "figure__img",
            "preserveAspectRatio": "none",
            "viewBox": "-1 -1 32 32",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "cx": "15",
                "cy": "15",
                "rx": "15",
                "ry": "15",
              },
              "children": Array [],
              "name": "ellipse",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "cx": undefined,
                "cy": undefined,
                "rx": undefined,
                "ry": undefined,
              },
              "x-attribsPrefix": Object {
                "cx": undefined,
                "cy": undefined,
                "rx": undefined,
                "ry": undefined,
              },
            },
          ],
          "name": "svg",
          "namespace": "http://www.w3.org/2000/svg",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "preserveAspectRatio": undefined,
            "viewBox": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "preserveAspectRatio": undefined,
            "viewBox": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "figure__dot _right-top",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "figure__dot _left-bot",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "figure__dot _left-top",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "figure__img",
          "preserveAspectRatio": "none",
          "viewBox": "-1 -1 32 32",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "cx": "15",
              "cy": "15",
              "rx": "15",
              "ry": "15",
            },
            "children": Array [],
            "name": "ellipse",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "cx": undefined,
              "cy": undefined,
              "rx": undefined,
              "ry": undefined,
            },
            "x-attribsPrefix": Object {
              "cx": undefined,
              "cy": undefined,
              "rx": undefined,
              "ry": undefined,
            },
          },
        ],
        "name": "svg",
        "namespace": "http://www.w3.org/2000/svg",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "figure__dot _right-bot",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "figure__dot _right-top",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "figure__dot _left-bot",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "figure__dot _left-top",
                },
                "children": Array [],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "preserveAspectRatio": undefined,
          "viewBox": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "preserveAspectRatio": undefined,
          "viewBox": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
